presto-main/src/main/java/com/facebook/presto/sql/planner/GroupingOperationRewriter.java

```diff
@@ -0,0 +1,91 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.sql.planner;
+
+import com.facebook.presto.metadata.Metadata;
+import com.facebook.presto.metadata.Signature;
+import com.facebook.presto.spi.type.Type;
+import com.facebook.presto.spi.type.TypeSignature;
+import com.facebook.presto.sql.analyzer.Analysis;
+import com.facebook.presto.sql.tree.Expression;
+import com.facebook.presto.sql.tree.ExpressionRewriter;
+import com.facebook.presto.sql.tree.ExpressionTreeRewriter;
+import com.facebook.presto.sql.tree.FunctionCall;
+import com.facebook.presto.sql.tree.GroupingOperation;
+import com.facebook.presto.sql.tree.QuerySpecification;
+import com.facebook.presto.type.ListLiteralType;
+
+import java.util.Arrays;
+import java.util.IdentityHashMap;
+import java.util.List;
+import java.util.stream.Collectors;
+
+import static com.facebook.presto.spi.type.BigintType.BIGINT;
+import static com.facebook.presto.spi.type.IntegerType.INTEGER;
+import static com.facebook.presto.sql.analyzer.ExpressionAnalyzer.resolveFunction;
+import static java.util.Objects.requireNonNull;
+
+public class GroupingOperationRewriter
+        extends ExpressionRewriter<Void>
+{
+    private final PlanBuilder subPlan;
+    private final QuerySpecification queryNode;
+    private final Analysis analysis;
+    private final Metadata metadata;
+    private final boolean isGroupingOperationTopLevel;
+
+    public GroupingOperationRewriter(PlanBuilder subPlan, QuerySpecification queryNode, Analysis analysis, Metadata metadata, boolean isGroupingOperationTopLevel)
+    {
+        this.subPlan = requireNonNull(subPlan, "subPlan is null");
+        this.queryNode = requireNonNull(queryNode, "node is null");
+        this.analysis = requireNonNull(analysis, "analysis is null");
+        this.metadata = requireNonNull(metadata, "metadata is null");
+        this.isGroupingOperationTopLevel = isGroupingOperationTopLevel;
+    }
+
+    @Override
+    public Expression rewriteExpression(Expression node, Void context, ExpressionTreeRewriter<Void> treeRewriter)
+    {
+        return treeRewriter.defaultRewrite(node, context);
+    }
+
+    @Override
+    public Expression rewriteGroupingOperation(GroupingOperation node, Void context, ExpressionTreeRewriter<Void> treeRewriter)
+    {
+        FunctionCall rewrittenExpression = subPlan.rewriteGroupingOperationToFunctionCall(node, queryNode);
+        if (isGroupingOperationTopLevel) {
+            analysis.setOutputExpressions(
+                    node,
+                    analysis.getOutputExpressions(queryNode).stream()
+                            .map(expression -> (expression instanceof GroupingOperation) ? rewrittenExpression : expression)
```

Comment:
Check: can outputs contain an output for a source that's also a GroupingOperation? If so, pandemonium will ensue.